README.md
Copyright 2016 Distributed Management Task Force, Inc. All rights reserved.
# Redfish Profile Simulator
## About
***Redfish Profile Simulator***
is a Python34 real simulator of the "simple monolithic server" feature profile.

  *  A simple, minimal Redfish Service
  *  For a monolithic Server
  *  Aligned with: OCP Remote Machine Management Spec feature set

### Description
* Based on flask
* Initial resources are loaded from a catfish mockup into python dictionary structures
  * After that, data is read/patched... to the dictionaries
* Supports BasicAuth, as well as Redfish Session Auth  (for one session, one user)
* Uses:
  * easy to add new URIs for testing a client
  * easy to tweak behavior or add bad responses to test a client
  * allows testing of authentication -- which current mockup servers dont do
  * easy to insert print statements in service to see if data coming across good..etc

### Current Limitation:
  * supports a single user/passwd and token
  * the user/passwd is:   root/password123456    
  * The authToken for Session Auth is: 123456SESSIONauthcode
  * Supports only HTTP  (not HTTPS)
  * with redfishtool, use options: redfishtool.py -r127.0.0.1:5000 -u root -p password123456 -S Never <subcmd>



## Usage
* ` python redfishProfileSimulatorMain.py [options]`

* `[Options]`:

		-V,  --Version,--- the program version
		-h,  --help,   --- help
		-H<hostIP>,  --Host=<hostIp>   --- host IP address. dflt=127.0.0.1
		-P<port>,--Port=<port> --- the port to use. dflt=5000
		-p<profile_path>, --profile=<profile_path>   --- the path to the Redfish profile to use. dflt="SimpleOcpServerV1"
    
    
## Implementation
* The simulation includes an http server, RestEngine, and dynamic Redfish datamodel.
* You can GET, PATCH,... to the service just like a real Redfish service.
* Both Basic and Redfish Session/Token authentication are supported 
  * for a single user/passwd and token
  * the user/passwd is:   root/password123456    
  * The authToken for Session Auth is: 123456SESSIONauthcode
  * these can be changed by editing the redfishURSs.py file---will make dynamic later.
* The http service and Rest engine is built on Flask, and all code is Python 3.4+
* The data model resources are "initialized" from the SPMF "SimpleOcpServerV1" Mockup.
  *  and stored as python dictionaries
  *  then the dictionaries are updated with patches, posts, deletes.
* The program can be extended to support other mockup \"profiles\".
* By default, the simulation runs on localhost (127.0.0.1), on port 5000.
  * These can be changed with CLI options: -P<port> -H <hostIP>  | --port=<port> --host=<hostIp>

## Simple OCP Server V1 Mockup Description
* A Monolithic server:
  * One ComputerSystem
  * One Chassis
  * One Manager

* Provides basic management features aligned with OCP Remote Machine Management Spec 1.01:
  * Power-on/off/reset
  * Boot to PXE, HDD, BIOS setup (boot override)
  * 4 temp sensors per DCMI (CPU1, CPU2, Board, Inlet)
  * Simple Power Reading, and  DCMI Power Limiting
  * Fan Monitoring w/ redundancy
  * Set asset tag and Indicator LED
  * Basic inventory (serial#, model, SKU, Vendor, BIOS ver…)
  * User Management
  * BMC management: get/set IP, version, enable/disable protocol

* What it does NOT have -- that the Redfish 1.0 model supports
  * No PSUs in model  (RMM spec did not include PSUs) 
  * No ProcessorInfo, MemoryInfo, StorageInfo, System-EthernetInterfaceInfo
  * No Tasks
  * JsonSchema and Registries collections left out (since that is optional)
  * No EventService--Remote Machine Management spec used basic PET alerts
  * Uses only the pre-defined privileges and roles


## TO DO
Some limitations to be extended in current implementation

* Auth supports a single hard-coded username, password, and AuthToken, although the protocol is 100% compliant with respect to testing clients trying to authenticate
  * ex with basic auth, you have to use the hard coded user/password
  * ex with Session Auth, you just use the hard coded AuthToken
* adding and deleting users not implemented--has 3 or 4 users predefined
* accountService properties can be written, but failed logins, lockouts, etc is not implemented
* system log not implemented yet

## Added SAS Fabric 
	* does exactly same function as DMTF Simple SAS fabric mockup example at http://redfish.dmtf.org/redfish/v1/mockup/847

## TO DO 
Add POST, PATCH and Delete functionality








CHANGELOG.md

# Change Log

## [0.9.2] - 2016-12-08
- fix code to work with late change in mockup data.  It was failing to load
- fixed incomplete code in systems.py modeling logs

## [0.9.1] - 2016-09-06
- Initial Public Release
- 

## [0.9.2] - 2017-10-11
- Add SAS Fabric support.


v1sim/fabrics.py
# Copyright Notice:
# Copyright 2016 Distributed Management Task Force, Inc. All rights reserved.
# License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/Redfish-Profile-Simulator/LICENSE.md

import os

from .resource import RfResource, RfCollection


class RfFabricsCollection(RfCollection):
    def element_type(self):
        return RfFabricObj

class RfFabricsCollection(RfCollection):
    def element_type(self):
        return RfFabricObj

#    def create_sub_objects(self, base_path, rel_path):
#        if os.path.isdir(os.path.join(base_path, rel_path)):
#            self.components["SAS"] = RfFabricObj(base_path,
#                                                 os.path.normpath("redfish/v1/Fabrics/SAS"),
#                                                              parent=self)


class RfFabricObj(RfResource):
    # create the dependent sub-objects that live under the Fabric object
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            if item == "Endpoints":
                self.components[item] = RfFabricEndpoints(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Switches":
                self.components[item] = RfFabricSwitchs(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Zones":
                self.components[item] = RfFabricZones(base_path, os.path.join(rel_path, item), parent=self)

    def patch_resource(self, patch_data):
        # first verify client didn't send us a property we cant patch
        for key in patch_data.keys():
            if key != "AssetTag" and key != "IndicatorLED":
                return 4, 400, "Invalid Patch Property Sent", ""
        # now patch the valid properties sent
        if "AssetTag" in patch_data:
            self.res_data['AssetTag'] = patch_data['AssetTag']
        if "IndicatorLED" in patch_data:
            self.res_data['IndicatorLED'] = patch_data['IndicatorLED']
        return 0, 204, None, None

    def reset_resource(self, reset_data):
        if "ResetType" in reset_data:
            # print("RESETDATA: {}".format(resetData))
            value = reset_data['ResetType']
            valid_values = self.res_data["Actions"]["#Chassis.Reset"]["ResetType@Redfish.AllowableValues"]
            if value in valid_values:
                # it is a supoported reset action  modify other properties appropritely
                if value == "On":
                    self.res_data["PowerState"] = "On"
                    print("PROFILE_SIM--SERVER WAS RESET. power now ON")
                elif value == "ForceOff":
                    self.res_data["PowerState"] = "Off"
                    print("PROFILE_SIM--SERVER WAS RESET. Power now Off")
                return 0, 204, "Chassis Reset", ""
            else:
                return 4, 400, "Invalid reset value: ResetType", ""
        else:  # invalid request
            return 4, 400, "Invalid request property", ""


# subclass Thermal Metrics
class RfFabricEndpoints(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            if item.__contains__("Drive"):
                self.components[item] = RfDriveCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)
            elif item.__contains__("Enclosure"):
                self.components[item] = RfEnclosureCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)
            elif item.__contains__("Initiator"):
                self.components[item] = RfInitiatorCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)
            if item.__contains__("TapeDrive"):
                self.components[item] = RfTapeDriveCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)

class RfFabricZones(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            self.components[item] = RfSwitchCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)


class RfFabricSwitchs(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            self.components[item] = RfSwitchCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)
class RfFabricPorts(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
                self.components[item] = RfPortCollection(base_path,
                                                          os.path.join(rel_path, item),
                                                          parent=self)
class RfSwitchCollection(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            if item == "Ports":
                self.components[item] = RfFabricPorts(base_path,
                                                  os.path.join(rel_path, item),
                                                  parent=self)


class RfPortCollection(RfResource):
    def element_type(self):
        return RfPortCollection

class RfDriveCollection(RfResource):
    def element_type(self):
        return RfDriveCollection

class RfEnclosureCollection(RfResource):
    def element_type(self):
        return RfEnclosureCollection

class RfInitiatorCollection(RfResource):
    def element_type(self):
        return RfInitiatorCollection

class RfTapeDriveCollection(RfResource):
    def element_type(self):
        return RfTapeDriveCollection



v1sim/serviceRoot.py
# Copyright Notice:
# Copyright 2016 Distributed Management Task Force, Inc. All rights reserved.
# License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/Redfish-Profile-Simulator/LICENSE.md

import os

from .accountService import RfAccountServiceObj
from .chassis import RfChassisCollection
from .fabrics import RfFabricsCollection
from .managers import RfManagersCollection
from .resource import RfResource, RfCollection
from .resource import RfResourceRaw
from .sessionService import RfSessionServiceObj
from .systems import RfSystemsCollection
from .updateService import RfUpdateServiceObj


class RfServiceRoot(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            if item == "odata":
                self.components[item] = RfOdataServiceDoc(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "$metadata":
                self.components[item] = RfOdataMetadata(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Systems":
                self.components[item] = RfSystemsCollection(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Chassis":
                self.components[item] = RfChassisCollection(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Managers":
                self.components[item] = RfManagersCollection(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Fabrics":
                self.components[item] = RfFabricsCollection(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "AccountService":
                self.components[item] = RfAccountServiceObj(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "SessionService":
                self.components[item] = RfSessionServiceObj(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "ResourceDirectory":
                self.components[item] = RfResourceDirectoryObj(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "UpdateService":
                self.components[item] = RfUpdateServiceObj(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "Registries":
                self.components[item] = RfRegistryCollection(base_path, os.path.join(rel_path, item), parent=self)
            elif item == "EventService":
                self.components[item] = RfEventServiceObj(base_path, os.path.join(rel_path, item), parent=self)

    def final_init_processing(self, base_path, rel_path):
        print("\n\n{}".format(self.res_data['Name']))


class RfOdataServiceDoc(RfResource):
    pass


class RfOdataMetadata(RfResourceRaw):
    pass


class RfResourceDirectoryObj(RfResource):
    pass


class RfRegistryCollection(RfCollection):
    def element_type(self):
        return RfRegistry


class RfRegistry(RfResource):
    pass


class RfEventServiceObj(RfResource):
    def create_sub_objects(self, base_path, rel_path):
        resource_path = os.path.join(base_path, rel_path);
        contents = os.listdir(resource_path)
        for item in contents:
            if item == "EventSubscriptions":
                self.components[item] = RfEventSubscriptionCollection(base_path,
                                                                      os.path.join(rel_path, item),
                                                                      parent=self)


class RfEventSubscriptionCollection(RfCollection):
    def element_type(self):
        return RfEventSubscription


class RfEventSubscription(RfResource):
    pass


redfishProfileSimulator.py
# Copyright Notice:
# Copyright 2016 Distributed Management Task Force, Inc. All rights reserved.
# License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/Redfish-Profile-Simulator/LICENSE.md

# This program is dependent on the following Python packages that should be installed separately with pip:
#    pip install Flask
#
# standard python packages
import sys
import getopt
import os

rfVersion = "0.9.49"
rfProgram1 = "redfishProfileSimulator"
rfProgram2 = "                "
rfUsage1 = "[-Vh]  [--Version][--help]"
rfUsage2 = "[-H<hostIP>] [-P<port>] [-p<profile_path>]"
rfUsage3 = "[--Host=<hostIP>] [--Port=<port>] [--profile_path=<profile_path>]"


def rf_usage():
        print("Usage:")
        print("  ", rfProgram1, "  ", rfUsage1)
        print("  ", rfProgram1, "  ", rfUsage2)
        print("  ", rfProgram2, "  ", rfUsage3)


def rf_help():
        print(rfProgram1,"implements a simulation of a redfish service for the \"Simple OCP Server V1\" Mockup.")
        print(" The simulation includes an http server, RestEngine, and dynamic Redfish datamodel.")
        print(" You can GET, PATHCH,... to the service just like a real Redfish service.")
        print(" Both Basic and Redfish Session/Token authentication is supported (for a single user/passwd and token")
        print("    the user/passwd is:   root/password123456.    The authToken is: 123456SESSIONauthcode")
        print("    these can be changed by editing the redfishURIs.py file.  will make dynamic later.")
        print(" The http service and Rest engine is built on Flask, and all code is Python 3.4+")
        print(" The data model resources are \"initialized\" from the SPMF \"SimpleOcpServerV1\" Mockup.")
        print("     and stored as python dictionaries--then the dictionaries are updated with patches, posts, deletes.")
        print(" The program can be extended to support other mockup \"profiles\".")
        print("")
        print(" By default, the simulation runs on localhost (127.0.0.1), on port 5000.")
        print(" These can be changed with CLI options: -P<port> -H <hostIP>  | --port=<port> --host=<hostIp>")
        print("")
        print("Version: ", rfVersion)
        rf_usage()
        print("")
        print("       -V,          --Version,                       --- the program version")
        print("       -h,          --help,                          --- help")
        print("       -H<hostIP>,  --Host=<hostIp>                  --- host IP address. dflt=127.0.0.1")
        print("       -P<port>,    --Port=<port>                    --- the port to use. dflt=5000")
        print("       -p<profile_path>, --profile=<profile_path>    --- the path to the Redfish profile to use. "
              "dflt=\"./MockupData/SimpleOcpServerV1\" ")


def main(argv):
    # set default option args
    #rf_profile_path = os.path.abspath("./MockupData/SimpleOcpServerV1")
    rf_profile_path = os.path.abspath("./MockupData/SimpleSASFabric")
    rf_host = "127.0.0.1"
    rf_port = 5000
    
    try:
        opts, args = getopt.getopt(argv[1:], "VhH:P:p:",
                                   ["Version", "help", "Host=", "Port=", "profile="])
    except getopt.GetoptError:
        print(rfProgram1, ":  Error parsing options")
        rf_usage()
        sys.exit(2)
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            rf_help()
            sys.exit(0)
        elif opt in ("-V", "--Version"):
            print("Version:", rfVersion)
            sys.exit(0)
        elif opt in ("-p", "--profile"):
            rf_profile_path = arg
        elif opt in "--Host=":
            rf_host = arg
        elif opt in "--Port=":
            rf_port=int(arg)
        else:
            print("  ", rfProgram1, ":  Error: unsupported option")
            rf_usage()
            sys.exit(2)

    print("{} Version: {}".format(rfProgram1,rfVersion))
    print("   Starting redfishProfileSimulator at:  hostIP={},  port={}".format(rf_host, rf_port))
    print("   Using Profile at {}".format(rf_profile_path))

    if os.path.isdir(rf_profile_path):
        # import the classes and code we run from main.
        from v1sim.serviceVersions import RfServiceVersions
        from v1sim.serviceRoot import RfServiceRoot
        # rfApi_SimpleServer is a function in ./RedfishProfileSim/redfishURIs.py.
        # It loads the flask APIs (URIs), and starts the flask service
        from v1sim.redfishURIs import rfApi_SimpleServer

        # create the root service resource
        root_path = os.path.normpath("redfish/v1")

        # create the version resource for GET /redfish
        versions = RfServiceVersions(rf_profile_path, "redfish")
        root = RfServiceRoot(rf_profile_path, root_path)

        # start the flask REST API service
        rfApi_SimpleServer(root, versions, host=rf_host, port=rf_port)
    else:
        print("invalid profile path")


if __name__ == "__main__":
    main(sys.argv)



    
    #http://127.0.0.1:5000/

    #app.run(host="0.0.0.0") # run on all IPs
    #run(host=None, port=None, debug=None, **options)
    #   host=0.0.0.0 server avail externally -- all IPs
    #   host=127.0.0.1 is default
    #   port=5000 default, or port defined in SERVER_NAME config var



